### PR TITLE
MinGW build support

### PR DESCRIPTION
**Pull request by hifi**
Adds universal MinGW/gcc build support for all targets.

This still needs work, only TD targets build now.

### CODE REVIEW

File: redalert/ftimer.h [hunk 1]
```diff
@@ -300,7 +300,7 @@ inline TTimerClass<T>::TTimerClass(unsigned long set)
 template <class T> inline unsigned long TTimerClass<T>::Value(void) const
 {
     unsigned long value = Accumulated;
-    if (Started != 0xFFFFFFFFU) {
+    if (this->Started != 0xFFFFFFFFU) {
```
Comment (OmniBlade) [inline redalert/ftimer.h]: Personally I think a better fix to the strict lookup rules is adding `using` directives to the class as seen here https://github.com/OmniBlade/Chronoshift/blob/b20acda5f8c20d8512bd916a51aa80828b0c186f/src/game/common/ttimer.h#L87. That way you have to touch less lines of code.

File: CMakeLists.txt [hunk 2]
```diff
@@ -2,10 +2,17 @@ cmake_minimum_required(VERSION 3.10)
 
 set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_SOURCE_DIR}/cmake/")
 
-enable_language(ASM_MASM)
-set(CMAKE_ASM_MASM_FLAGS "/Zm /Cp /safeseh" CACHE INTERNAL "MASM flags")
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /Zp1")
-add_definitions(-DWINDOWS_IGNORE_PACKING_MISMATCH)
+if(WIN32)
```
Comment (OmniBlade) [inline CMakeLists.txt:5]: Should probably be gated on MSVC, not WIN32 which I assume is still set for mingw.

File: redalert/function.h [hunk 3]
```diff
@@ -109,6 +109,18 @@ Map(screen) class heirarchy.
 
 #define _MAX_NAME _MAX_FNAME
 
+#ifdef __MINGW32__
```
Comment (OmniBlade) [inline redalert/function.h]: Could add the NOMINMAX macro and define it for all targets unconditionally? Ideally we move to std::min/max down the road anyhow for the C++ code.

File: tiberiandawn/vector.h [hunk 4]
```diff
@@ -873,7 +873,7 @@ template <class T> int DynamicVectorClass<T>::ID(T const& object)
 template <class T> int DynamicVectorClass<T>::Add(T const& object)
 {
     if ((unsigned)ActiveCount >= Length()) {
-        if ((IsAllocated || !VectorMax) && GrowthStep > 0) {
+        if ((this->IsAllocated || !this->VectorMax) && GrowthStep > 0) {
```
Comment (OmniBlade) [inline tiberiandawn/vector.h:876]: Another case where we could use `using` directives in the class?

**Review by OmniBlade — CHANGES_REQUESTED** [review]
I realise this is WIP, just a few things I noticed that you might or might not already have on your radar.

File: CMakeLists.txt (hunk 2 above)
Comment (hifi) [reply CMakeLists.txt:5]: Yeah, that should be fixed now

File: redalert/function.h (hunk 3 above)
Comment (hifi) [reply redalert/function.h]: Fine for me, need to do that

File: CMakeLists.txt (hunk 2 above)
Comment (hifi) [reply CMakeLists.txt:5]: That's actually wrong, `WIN32` is not defined for MinGW so this works. We don't yet know our compiler at that point so `MSVC` is not defined as it's based on compiler detection.

Need to dig a bit further.

File: tiberiandawn/vector.h (hunk 4 above)
Comment (hifi) [reply tiberiandawn/vector.h:876]: This didn't actually work in my naive attempt in this case. It caused inheritance issues.

File: CMakeLists.txt (hunk 2 above)
Comment (hifi) [reply CMakeLists.txt:5]: One solution would be to gate just the assembler stuff behind `WIN32` and set the other stuff after the compiler has been detected with `MSVC`?

**Comment by hifi** [comment]
Remastered RA is still grayscale on MinGW, uh oh.

**Review by OmniBlade — APPROVED** [review]

**Comment by hifi** [comment]
Narrowed this down to fixed 1/2 not being initialized before OptionsClass constructor. Fixing this seems to require writing constexpr initializers for static helpers.

Not sure if it's just lucky on MSVC the fixed type gets initialized first or if there's a guaranteed static initialization order.

**Pull request merged**